### DISCUSSION TITLE
Sources and Sinks

### DISCUSSION THREAD

**Discussion by alepping** [nebulastream/nebulastream#8] [Ideas]
# Description
This discussion is about how to 'best' support sources and sinks. 
## Goal
Develop a concrete plan for how to enable sources and sinks that:
- require less classes / programming logic in the repository (code refactor)
- that can be 'easily' extended by implementing a 'sink/source' plugin

## Problem
**Implementation:**
Currently, to create a new sink and source in NebulaStream, we need to edit 16+ files across the codebase (not in one place), ranging from the actual source and sink implementations to the descriptors, to the lowering logic to the serialization. This makes it very cumbersome to add new sources and sinks which is a **major** hindrance to adopting NebulaStream for new use cases (community adoption (can be research community!)).

## Ideas
**Implementation Sources:**
- use a source registry on the worker
- use the logical source name to read from the registry and return a correctly configured physical source (no descriptor needed) 
- have a single source interface (DataSource), which means we can also return all different physical sources using the same abstract representation.
- for serialization load the parameters from the source/sink registry on the worker using the source/sink name (essentially the parameters that are now contained in the descriptor and pass it to the serialization
- for deserialization, we register the source/sink in the respective registry using the corresponding name

This design could make it possible to develop a source registry that would enable new users to relatively simply create new sources.

**Implementation Sinks:**
Currently, we attach sinks to queries. A query must have a SinkDescriptor, in which we construct a specific sink description with all necessary parameters. It might be an interesting idea to handle sinks similar to sources. In particular, we could create sinks using 'DDL' syntax outside of queries and give them names (similar to table names). This would allow us to reuse sinks that we already created. Additionally, queries would become as simple as using the correct sink (and source) name in the query.

## Solution
After investigating sources and sinks in the current NebulaStream implementation, discussing the topic with @ls-1801 and @keyseven123 and based on the suggestions made by @PhilippGrulich in #10, especially the comment 'the worker is stateless and keeps no state of physical sources.' We came to the following conclusion.

For the first milestone:
- all queries that arrive at the worker contain source/sink descriptors (that fully describe a physical source/sink ((meta information), type and configuration)prepares to run a query that requires a specific physical source/sink
- the worker has a source and a sink registry
- the worker can load a source/sink implementation from the registry given a source name (or source type) supplied with the source/sink descriptors
- the source/sink descriptors (usually) contain configurations for the source/sink that the worker can use to configure the source/sink
- thus, the worker does not keep track of any state concerning physical sources and sinks, it only creates physical sources and sinks when it prepares to run a specific decomposed query plan

Assumptions:
- the coordinator keeps track of which physical sources belong to which logical sources and which workers support the physical sources
- in the future, it will be possible to define a physical source on the coordinator for a specific worker. Additionally, it will be possible to bind physical sources to logical sources (possibly in a separate step from defining the physical source)
- the coordinator maps a logical source to physical source descriptors that are then included in the query that is sent to each worker

**Comment by keyseven123** [comment]
Hi,
couple comments from my side. In general, your points are valid and I agree with them. Regarding the problems you specified. Some of them are not source/sink specific but general deployment, or am I mistaken?

I am not quite sure how it is different to our current design for the sources. We currently already have a source registry. 

I like your implementation idea of the sinks.

**Reply by alepping** [reply]
I edited the main message discussion. Hopefully, your concerns are addressed now.

**Reply by PhilippGrulich** [reply]
Sounds like a good idea. One thing to keep in mind would be how we can handle the case if multiple queries writing to the same sink? For example a file?

**Reply by zeuchste** [reply]
Hi in general this sound good but there are some important points missing to understand it fully:

1. When you say source registry where is this on a node or the coordinator? 
2. The problem with sources is that they are totally node specific. Imagine a TPC source or a CSV source the logical source is easy to maintain globally but the physical source is different for each node, i.e., different ip and different file path
3. It is unclear when exactly the sources are registered and where. So at startup over the coordinator via rest or on node startup on the command line or on node setup via config file
4. One major problem that we always had with sources is that users where overwhelmed with the parameters they have to specify, here we should pick very good defaults (like every second), document them, and only let the user specify the bare minimum
5. A different idea would be to turn around the source registration. So instead of the worker starts and says this are my sources to the coordinator, the coordinator sends a list of sources and the worker decides which one he can offer (fix idea not well thought) 
6. In general we need so separate here the use of the source (cumbersome over config files) and the addition of new sources (cumbersome as we have to attach many files). Both should be considered

**Reply by PhilippGrulich** [reply]
While reading I had the following idea. Maybe it would be easier to manage both logical and physical sources at the coordinator.
As a result:
- all queries are bound to specific physical sources when they arrive at the worker.
- the worker is stateless and keeps no state of physical sources. 
- the coordinator is the only entity, which manages the mapping from logical to physical sources.
- we could have two ways to register a new physical source:
1. via the worker configuration -> basicall the worker just tells the coordinator which sources it has.
2. via the coordinator -> for example using some DML statement: 
**BIND FILE("PATH") AT "NODE2" TO STREAM "X"** 

This would also have the benefit that we could support "flexible" physical sources. 
For example, some sources like KAFKA could be placed on any node.

**Reply by ls-1801** [reply]
I like @zeuchste idea for the coordinator reaches out to all worker about which physical sources it supports.

But for some sources (actually almost all) the source type is not enough and whether a specific source can be instantiated on a worker depends on the configuration. E.g a filesource may require a specific file to exist.

However I think it would also make sense to ask for support, checking if plugin is loaded, and creating an Error when trying to start a query with specific configuration.

For our first milestone I would keep it simple and just accept arbitrary source configuration and report an error if the source cannot be established, I think the plugin system and solid data source implementation is more important

**Reply by zeuchste** [reply]
@PhilippGrulich  yes your approach is also valid and good. But again, currently we do have the assumption that outside world do not know anything on inside world. If you let the binding happen over the coordinator, this is violated (which is maybe not bad but we need then to accept this as a new property). 

The stateless idea of physical config is maybe not possible as the worker has to start and stop and interact with the source locally, how will he do it if he does not know about it.

After reading the entire discussion, I think the main thing we need to change is that we allow DML like BIND over the coordinator for physical sources which is currently not possible. This would enable us to for example use a setup script like @alepping suggested.

@ls-1801 thanks for supporting the idea of a coordinator ask the sources what they can contribute. This would be really cool but again how is this organized how can a worker decide that yes he can open a TCP port 5555 but if he wants to. Also the files can lie on different path maybe. If anybody has a cool idea to solve this, I think this would be the best solution.

**Reply by alepping** [reply]
@PhilippGrulich , @zeuchste and @ls-1801 , I moved your honestly great (no /s) comments to discussion #10.
This discussion should focus on the actual implementation of the sources and sinks on the worker side. Discussion #10 is the high-level vision of sources and sinks in the system.

**Comment by klauck** [comment]
Hi,
thanks for starting the discussion.
I still have to understand how sources and sinks work (i.e., are currently implemented), maybe starting with sources and what have to be implemented to add a new source.

From my current understanding, a tuple buffer is the "interface" for operators,
but I already do not know how the filling is triggered (e.g., whether this depends on the specific data source).

What are pointers to read about this and things to be implemented for a new source?
I already saw the source implementations in LogicalOperators and nes-runtime.
